### DISCUSSION TITLE
Configuration wizard taking long time to start up

### DISCUSSION THREAD

**Discussion by Carl-mgb66** [nathanmarlor/foxess_modbus#384] [Support]
Hi,

For a variety of reasons I removed and re-added this integration. Now I get a window saying "Please wait, starting configuration wizard for foxess_modbus" when I try to add the integration. Any thoughts on why this is happening?

**Comment by canton7** [comment]
Normally it means that HA is installing things behind the scenes. Look at the log (Settings - System - Logs) and see what's going on.

**Reply by Carl-mgb66** [reply]
Seems to be working now - however it can't connect to my inverter which was part of the problem I was trying to solve. Connected via LAN, was working before, but recently seems to have fallen over more often.

**Reply by canton7** [reply]
What's the exact error? You should also see relevant messages in the log, which would be useful.

**Reply by Carl-mgb66** [reply]
I get the following in the log:

Logger: custom_components.foxess_modbus.modbus_controller
Source: custom_components/foxess_modbus/modbus_client.py:289
Integration: FoxESS - Modbus ([documentation](https://github.com/nathanmarlor/foxess_modbus), [issues](https://github.com/nathanmarlor/foxess_modbus/issues))
First occurred: 17:40:15 (2 occurrences)
Last logged: 17:40:19

Autodetect: failed to connect to (tcp://192.168.50.228:502)
Traceback (most recent call last):
  File "/config/custom_components/foxess_modbus/modbus_controller.py", line 331, in autodetect
    await client.read_registers(
  File "/config/custom_components/foxess_modbus/modbus_client.py", line 198, in read_registers
    response = await self._async_pymodbus_call(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/foxess_modbus/modbus_client.py", line 289, in _async_pymodbus_call
    result = await self._hass.async_add_executor_job(call, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pymodbus/client/mixin.py", line 102, in read_holding_registers
    return self.execute(
           ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pymodbus/client/base.py", line 179, in execute
    raise ConnectionException(f"Failed to connect[{self!s}]")
pymodbus.exceptions.ConnectionException: Modbus Error: [Connection] Failed to connect[ModbusTcpClient(192.168.50.228:502)]

It's worked before, seemed to stop working with more recent updates.

Firmware: 
Master: 1.48
Slave: 1.02
Manager: 1.45

I can ping the inverter and a port scan shows 502 as open.

**Reply by canton7** [reply]
Is that a dynamic ip, or have you statically assigned it? Is there any chance it's collided with the ip of another device on your network?

**Reply by Carl-mgb66** [reply]
It's a statically assigned IP - set up in the DHCP server. I did try setting a fixed IP in the inverter - that made no difference.

**Reply by canton7** [reply]
> seemed to stop working with more recent updates.

Are you referring to inverter updates or integration updates? There haven't been any updates to the integration which should affect the ability to open a tcp connection, which seems to be what's happening here.

**Reply by Carl-mgb66** [reply]
It’s been an issue following some recent HA updates and integration updates. I found increasingly that following an update it wouldn’t connect and I’d have to reboot the whole HA machine (a raspberry pi). Now it just doesn’t seem to want to connect at all.

No inverter updates.

If I try to add the inverter now, I’m getting a blank window rather than the Fox config window. Looking in the logs I’m seeing..

Logger: frontend.js.latest.202307051
Source: components/system_log/__init__.py:257 
First occurred: 09:56:21 (1 occurrences) 
Last logged: 09:56:21

:0:0 Script error.

Any thoughts?

**Reply by canton7** [reply]
When it's having trouble connecting, can you shell into the raspberry pi and see if you can use netcat / modpoll to establish a connection on port 503? 

If the UI it having trouble loading, a ctrl-F5 refresh normally sorts it out for me.

**Reply by Carl-mgb66** [reply]
ctrl-F5 doesn't seem to be working - I'm getting the screen below:

![Screenshot 2023-08-01 at 20 28 50](https://github.com/nathanmarlor/foxess_modbus/assets/75570551/bd96937e-6400-47d0-8d36-b4baf366e0ec)

I couldn't find netcat or modpoll on the pi, but managed to run netcat from my Mac - it connected to port 502 and downloaded a modbus app which managed to connect but reported invalid data - see below:

![Screenshot 2023-08-01 at 20 41 46](https://github.com/nathanmarlor/foxess_modbus/assets/75570551/a401153a-cfb3-44f1-ba38-58a28e4b8180)

Don't know if this helps at all.

**Reply by canton7** [reply]
I'm particularly interested in whether your pi can establish a connection, since that's the one that's having trouble.

I'm not sure what os your pi is running, but you should be able to get some basic networking tools on there? Modpoll should be a straight download fe the Web.

Might also be interesting to see how many fds are used? Although I'd expect that to cause a different error.

Not sure on the dialog not rendering, but my experience is that issues like that tend to be browser cache related. Is there nothing in your browser's console?

**Reply by Carl-mgb66** [reply]
My pi is running the standard Home Assistant Operating System as downloaded from the Home Assistant site.

Not sure what I'd be looking for in the browser's console - doesn't seem to be anything there that I can see. FYI - I tried a completely different browser and other devices, same effect, even on the phone app.

**Reply by canton7** [reply]
Anything in red really. Just copy/paste in here if you're not sure. Also check the networking tab for any failures?

I'm thinking at the moment that maybe there's a problem with your pi's connectivity: that would also explain why bits of the UI aren't loading.

Maybe also do a `ping -t <pi's ip>` from your mac, see if you get any drops?

**Reply by Carl-mgb66** [reply]
In terms of the browser issue - I'm seeing the following errors:

![Screenshot 2023-08-02 at 13 48 16](https://github.com/nathanmarlor/foxess_modbus/assets/75570551/788b6388-c80e-4d9b-a181-e407bcac51e3)

And the following warnings:


![Screenshot 2023-08-02 at 13 48 34](https://github.com/nathanmarlor/foxess_modbus/assets/75570551/42c84da8-17f6-4ff8-b9b6-c6ed76e7281b)

Nothing on the networking tab.

Ping from mac to HA (on the pi) - no packet loss

![Screenshot 2023-08-02 at 13 53 45](https://github.com/nathanmarlor/foxess_modbus/assets/75570551/89adcb13-affd-45d9-8663-74910d73bdc0)

I managed to run netcat on the Home Assistant (pi) and got the following:

![Screenshot 2023-08-02 at 14 04 01](https://github.com/nathanmarlor/foxess_modbus/assets/75570551/31a0c322-b986-4a3d-b3d6-2f53c6e4411e)

**Reply by canton7** [reply]
I'm slightly surprised at those ping times -- they're significantly higher than I would expect for a local network. Is your raspberry pi under heavy load or something?

Did you check for browser / network errors after attempting (and failing) to open the config wizard?

I'd be interesting in seeing pings over a longer period of time, see if you have drop-outs. I'd also be interesting in seeing the pi succeeding in pinging / nc-ing to the inverter at the same time that the integration fails to connect.

**Reply by Carl-mgb66** [reply]
The screenshots above are the errors in the browser after attempting (and failing) to open the config wizard.

I ran the ping from the mac to the pi through 50 entries, no packet loss, avg 8.9ms ping time. The pi is connected by cable to an ASUS Zen Wifi adapter which is the slave in my mesh network. The mac is on the end of a Devolo powerline adapter which links to the master in my mesh network.

I can't do the ping/nc-ing from the pi to the inverter at the same time as the integration fails to connect as I can't get a config window at present.

**Reply by canton7** [reply]
When you (try to) open the config dialog, you should see some network requests go out on the Network tab. See what's happening to those?

Also:

>  Is your raspberry pi under heavy load or something?

> Might also be interesting to see how many fds are used? Although I'd expect that to cause a different error.

**Reply by Carl-mgb66** [reply]
Hi, firstly thanks for the help.

I've tried opening the config dialog with the console and network tabs open, see the result below (I cleared the tabs before executing ADD INTEGRATION etc.)

![Screenshot 2023-08-02 at 20 18 36](https://github.com/nathanmarlor/foxess_modbus/assets/75570551/b5f06947-723b-465e-9a0c-31dd623c63e4)

Nothing in the console and the last entry in the network tab is shown.

> Is your raspberry pi under heavy load or something?

Not particularly - currently running at around 6% processor

> Might also be interesting to see how many fds are used? Although I'd expect that to cause a different error.

Not entirely sure what you mean by "how many fds are used?" If you can point me in the right direction I'll check.

**Reply by canton7** [reply]
>Not entirely sure what you mean by "how many fds are used?" If you can point me in the right direction I'll check.

E.g. `lsof | wc -l`

> Nothing in the console and the last entry in the network tab is shown.

Can you open any other "Configure" dialogs from other integrations? Is anything written to the system log when the dialog fails to load properly?

**Reply by Carl-mgb66** [reply]
Well, that's strange...

It's all working fine now - I've just applied all the various updates to the HA including new CORE and OS updates and update all my integrations that were pending whilst I've been away, and I tried the config again. After a moments delay it all worked and everything is now working fine.

Out of interest I did run:

`lsof | wc -l`

And got a response of 95 - don't know if this is unusual or not.

FYI - I could open other "Configure" dialogs.

As I say - all seems to be working find now. Thanks for your input.

**Reply by canton7** [reply]
Cool, thanks for letting me know!

95 is fine, nowhere near the limits.

**Reply by Carl-mgb66** [reply]
Sadly, spoke too soon. The connection to the inverter has stopped working again.

The good news I can select Configure from the integration screen and go through the config screens (without deleting the current config) and it just says:

![Screenshot 2023-08-15 at 19 55 37](https://github.com/nathanmarlor/foxess_modbus/assets/75570551/a0317a8d-03cf-41b9-8245-a6ac3a860ab0)

The system log says:

`Logger: custom_components.foxess_modbus.modbus_controller
Source: custom_components/foxess_modbus/modbus_client.py:289
Integration: FoxESS - Modbus ([documentation](https://github.com/nathanmarlor/foxess_modbus), [issues](https://github.com/nathanmarlor/foxess_modbus/issues))
First occurred: 18:30:11 (3 occurrences)
Last logged: 19:53:45

Autodetect: failed to connect to (tcp://192.168.50.228:502)
Traceback (most recent call last):
  File "/config/custom_components/foxess_modbus/modbus_controller.py", line 331, in autodetect
    await client.read_registers(
  File "/config/custom_components/foxess_modbus/modbus_client.py", line 198, in read_registers
    response = await self._async_pymodbus_call(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/foxess_modbus/modbus_client.py", line 289, in _async_pymodbus_call
    result = await self._hass.async_add_executor_job(call, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pymodbus/client/mixin.py", line 102, in read_holding_registers
    return self.execute(
           ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pymodbus/client/base.py", line 179, in execute
    raise ConnectionException(f"Failed to connect[{self!s}]")
pymodbus.exceptions.ConnectionException: Modbus Error: [Connection] Failed to connect[ModbusTcpClient(192.168.50.228:502)]`

I had debug logging switched on and that shows:

`2023-08-15 19:53:45.335 ERROR (MainThread) [custom_components.foxess_modbus.modbus_controller] Autodetect: failed to connect to (tcp://192.168.50.228:502)`
`Traceback (most recent call last):`
`  File "/config/custom_components/foxess_modbus/modbus_controller.py", line 331, in autodetect
    await client.read_registers(`
`  File "/config/custom_components/foxess_modbus/modbus_client.py", line 198, in read_registers
    response = await self._async_pymodbus_call(`
`  File "/config/custom_components/foxess_modbus/modbus_client.py", line 289, in _async_pymodbus_call
    result = await self._hass.async_add_executor_job(call, *args)`
`  File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)`
`  File "/usr/local/lib/python3.11/site-packages/pymodbus/client/mixin.py", line 102, in read_holding_registers
    return self.execute(`
`  File "/usr/local/lib/python3.11/site-packages/pymodbus/client/base.py", line 179, in execute
    raise ConnectionException(f"Failed to connect[{self!s}]")`
`pymodbus.exceptions.ConnectionException: Modbus Error: [Connection] Failed to connect[ModbusTcpClient(192.168.50.228:502)]`
`2023-08-15 19:53:45.344 DEBUG (MainThread) [custom_components.foxess_modbus.modbus_client] Closing connection to modbus on tcp://192.168.50.228:502`
`2023-08-15 19:53:45.803 DEBUG (MainThread) [custom_components.foxess_modbus.modbus_controller] Reading addresses on tcp://192.168.50.228:502 247: (31000, 39)`
`2023-08-15 19:53:45.806 DEBUG (MainThread) [custom_components.foxess_modbus.modbus_controller] Failed to connect to tcp://192.168.50.228:502 247: Modbus Error: [Connection] Failed to connect[ModbusTcpClient(192.168.50.228:502)]`
`2023-08-15 19:53:55.803 DEBUG (MainThread) [custom_components.foxess_modbus.modbus_controller] Reading addresses on tcp://192.168.50.228:502 247: (31000, 39)`
`2023-08-15 19:53:55.806 DEBUG (MainThread) [custom_components.foxess_modbus.modbus_controller] Failed to connect to tcp://192.168.50.228:502 247: Modbus Error: [Connection] Failed to connect[ModbusTcpClient(192.168.50.228:502)]`

Doing a nc in Terminal on the HA gives:

`nc -zv 192.168.50.228 502`
`192.168.50.228 (192.168.50.228:502) open`

The connection appeared to fail when I restarted the HA after updating a couple of other integrations, but not this one.

**Reply by canton7** [reply]
Can you just run `pip show pymodbus` on the HA machine?

**Reply by canton7** [reply]
I think HA has upgraded its dependency on pymodbus, and that's broken the way in which it triggers a connection. However, we still rely on the old pymodbus version. So if you have two integrations which both depend on pymodbus, and the other integration depends on the newer version, then as far as I can work out it's random as to which one gets installed. Which probably explains the intermittent behaviour here.

**Reply by canton7** [reply]
Can you try the pre-release [Version 1.7.1b1](https://github.com/nathanmarlor/foxess_modbus/releases/tag/v1.7.1b1)?

**Reply by Carl-mgb66** [reply]
> Can you try the pre-release [Version 1.7.1b1](https://github.com/nathanmarlor/foxess_modbus/releases/tag/v1.7.1b1)?

Hi, I tried this and get an error:

Error connecting to your inverter. Details: '_params' object has no attribute 'timeout'

I suspect you may be on to something with the pymodbus version as I have another Solaredge integration that uses modbus. I'll run pip when I'm able, I'm at work now.

**Reply by canton7** [reply]
OK, thanks. Try v1.7.1b2. I'm also at work, so having to test this without actually being able to connect to an inverter...

**Reply by Carl-mgb66** [reply]
Thanks, that seems to have worked. Appreciate your support.

**Reply by canton7** [reply]
Thanks for confirming!
